File: Varios/Programs/amigo_invisible.java
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.ArrayList;

public class amigo_invisible {

    public static void main(String[] args) {    

        getData();     
    }

    private static int getNumParticipants(Scanner scn) {

        System.out.println("Number of participants: ");
        int participantsNum = scn.nextInt();scn.nextLine();
        
        return participantsNum;
    }

    private static void getData() {

        Scanner scn = new Scanner(System.in);

        int participantsNum = getNumParticipants(scn);

        ArrayList<String> participantsList = new ArrayList<>();

        for (int i = 0; i < participantsNum; i++) {
            System.out.println("Name participant " + (i+1) + ":");
            participantsList.add(scn.nextLine());
        }
        Collections.shuffle(participantsList);
        scn.close();

        calculateResult(participantsList, participantsNum);
    }

    private static void calculateResult(ArrayList<String> participantsList, int participantsNum) {

        Map<String,String> secretSantas = new HashMap<>();

        for (int i = 0; i < participantsNum; i++) {
            String secretSanta = participantsList.get((i+1) % participantsNum);
            secretSantas.put(participantsList.get(i), secretSanta);
        }
        showResult(secretSantas);
    }

    private static void showResult(Map<String,String> secretMap) {

        System.out.println("SECRET SANTA RESULT");
        for (Map.Entry<String, String> ss : secretMap.entrySet()) {
            System.out.println(ss.getKey() + "-->" + ss.getValue());
        }
    }
}

File: Varios/Programs/amigo_invisible.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	
	var participantsNum int
	fmt.Print("Number of participants: ")
	fmt.Scan(&participantsNum)

	var participants []string

	for i := 0; i < participantsNum; i++ {
		var name string
		fmt.Printf("Name participant %d: ", i+1)
		fmt.Scan(&name)
		participants = append(participants, name)
	}
	
	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(participants), func(i, j int) {
		participants[i], participants[j] = participants[j], participants[i]
	})

	secretSantas := make(map[string]string)

	for i := 0; i < participantsNum; i++ {
		secretSanta := participants[(i+1)%participantsNum]
		secretSantas[participants[i]] = secretSanta
	}

	fmt.Println("\nSECRET SANTA RESULT")
	for key, value := range secretSantas {
		fmt.Printf("%s --> %s\n", key, value)
	}
}
